src/deluge/storage/storage_manager.cpp

```diff
@@ -809,61 +836,90 @@ void FileReader::readDone() {
 	}
 }
 
-FRESULT FileReader::closeFIL() {
-	return f_close(&readFIL);
+FRESULT FileReader::closeWriter() {
+	if (!memoryBased)
+		return f_close(&readFIL);
+	else
+		return FRESULT::FR_OK;
 }
 
 FileWriter::FileWriter() {
-	void* temp = GeneralMemoryAllocator::get().allocLowSpeed(32768 + CACHE_LINE_SIZE * 2);
+	bufferSize = 32768;
+	void* temp = GeneralMemoryAllocator::get().allocLowSpeed(bufferSize + CACHE_LINE_SIZE * 2);
 	writeClusterBuffer = (char*)temp + CACHE_LINE_SIZE;
 }
 
+FileWriter::FileWriter(bool inMem) : FileWriter() {
+	memoryBased = true;
+}
+
 FileWriter::~FileWriter() {
 	GeneralMemoryAllocator::get().dealloc(writeClusterBuffer);
 }
 
+int32_t FileWriter::bytesWritten() {
+	return fileTotalBytesWritten + fileWriteBufferCurrentPos;
+}
+
 void FileWriter::resetWriter() {
 	fileWriteBufferCurrentPos = 0;
 	fileTotalBytesWritten = 0;
 	fileAccessFailedDuringWrite = false;
 }
 
-FRESULT FileWriter::closeFIL() {
+FRESULT FileWriter::closeWriter() {
+	if (memoryBased) {
+		if (fileWriteBufferCurrentPos < bufferSize) {
+			writeClusterBuffer[fileWriteBufferCurrentPos] = 0;
+			return FRESULT::FR_OK;
+		}
+		else {
+			return FRESULT::FR_INT_ERR;
+		}
+	}
 	return f_close(&writeFIL);
 }
 
-void FileWriter::writeChars(char const* output) {
-	while (*output) {
-		if (fileWriteBufferCurrentPos == audioFileManager.clusterSize) {
-			if (!fileAccessFailedDuringWrite) {
-				Error error = writeBufferToFile();
-				if (error != Error::NONE) {
-					fileAccessFailedDuringWrite = true;
-					return;
-				}
-			}
-
-			fileWriteBufferCurrentPos = 0;
-		}
-
-		writeClusterBuffer[fileWriteBufferCurrentPos] = *output;
-
-		output++;
-		fileWriteBufferCurrentPos++;
+void FileWriter::writeBlock(uint8_t* block, uint32_t size) {
+	for (uint32_t ix = 0; ix < size; ++ix) {
+		writeByte(block[ix]);
+	}
+}
 
-		// Ensure we do some of the audio routine once in a while
-		if (!(fileWriteBufferCurrentPos & 0b11111111)) {
-			AudioEngine::logAction("writeCharsJson");
+void FileWriter::writeByte(int8_t b) {
 
-			// AudioEngine::routineWithClusterLoading();
+	if (fileWriteBufferCurrentPos == bufferSize) {
+		if (!memoryBased && !fileAccessFailedDuringWrite) {
+			Error error = writeBufferToFile();
+			if (error != Error::NONE) {
+				fileAccessFailedDuringWrite = true;
+				return;
+			}
+		}
+		if (memoryBased) {
+			fileAccessFailedDuringWrite = true;
+			return;
+		}
+		fileWriteBufferCurrentPos = 0;
+	}
 
-			uiTimerManager.routine();
+	writeClusterBuffer[fileWriteBufferCurrentPos] = b;
+	fileWriteBufferCurrentPos++;
 
-			if (display->haveOLED()) {
-				oledRoutine();
-			}
-			PIC::flush();
+	// Ensure we do some of the audio routine once in a while
+	if (callRoutines && !(fileWriteBufferCurrentPos & 0b11111111)) {
```

Comment:
this should probably call routineForSD instead of just the oled and pad stuff, dunno why it wasn't before

Reply:
I will put in routineForSD and hope nothing explodes.